test/Microsoft.SqlTools.ServiceLayer.UnitTests/QueryExecution/DataStorage/SaveAsMarkdownFileStreamWriterTests.cs

```diff
@@ -0,0 +1,383 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.SqlTools.ServiceLayer.QueryExecution.Contracts;
+using Microsoft.SqlTools.ServiceLayer.QueryExecution.DataStorage;
+using Microsoft.SqlTools.ServiceLayer.UnitTests.Utility;
+using NUnit.Framework;
+
+namespace Microsoft.SqlTools.ServiceLayer.UnitTests.QueryExecution.DataStorage
+{
+    public class SaveAsMarkdownFileStreamWriterTests
+    {
+        // Regex: Matches '|' not preceded by a '\'
+        private static readonly Regex UnescapedPipe = new Regex(@"(?<!\\)\|", RegexOptions.Compiled);
+
+        [Test]
+        public void Constructor_NullStream()
+        {
+            // Act
+            TestDelegate action = () => _ = new SaveAsMarkdownFileStreamWriter(
+                null,
+                new SaveResultsAsMarkdownRequestParams(),
+                Array.Empty<DbColumnWrapper>()
+            );
+
+            // Assert
+            Assert.Throws<ArgumentNullException>(action);
+        }
+
+        [Test]
+        public void Constructor_NullColumns()
+        {
+            // Act
+            TestDelegate action = () => _ = new SaveAsMarkdownFileStreamWriter(
+                Stream.Null,
+                new SaveResultsAsMarkdownRequestParams(),
+                null
+            );
+
+            // Assert
+            Assert.Throws<ArgumentNullException>(action);
+        }
+
+        [Test]
+        public void Constructor_WithoutSelectionWithHeader_WritesHeaderWithAllColumns()
+        {
+            // Setup:
+            // ... Create a request params that has no selection made, headers should be printed
+            // ... Create a set of columns
+            // --- Create a memory location to store the output
+            var requestParams = new SaveResultsAsMarkdownRequestParams { IncludeHeaders = true };
+            var (columns, _) = GetTestValues(2);
+            byte[] output = new byte[8192];
+            using var outputStream = new MemoryStream(output);
+
+            // If: I construct a Markdown file writer
+            var writer = new SaveAsMarkdownFileStreamWriter(outputStream, requestParams, columns);
+            writer.Dispose();
```

Comment:
`using (var writer...)`

Here and elsewhere

Reply:
Originally I wasn't doing that because I wanted to make sure disposal didn't do anything weird. Since `using var xyz;` (I think) disposes on method exit and `using (var writer) { }` would have this pointless empty block. But, I guess since there isn't anything fancy written in `Dispose`, I guess we can just use the `using var xyz;`.